In [1]:
# Import libraries
import pandas as pd

# Read the CSV and Perform Basic Data Cleaning

In [2]:
# Read CSV file 
df = pd.read_csv("exoplanet_data.csv")

# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')

# Drop the null rows
df = df.dropna()
df.head(10)

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   
5       CONFIRMED              0              0              0              0   
6       CONFIRMED              0              0              0              0   
7       CONFIRMED              0              0              0              0   
8       CONFIRMED              0              1              0              0   
9       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   
5    2.566589     1.780000e-05    -1.780000e-05   179.554370   
6   16.068647     1.090000e-05    -1.090000e-05   173.621937   
7    2.470613     2.700000e-08    -2.700000e-08   122.763305   
8    2.204735     4.300000e-08    -4.300000e-08   121.358542   
9    3.522498     1.980000e-07    -1.980000e-07   121.119423   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   
5          0.004610  ...            -232      4.486           0.054   
6          0.000517  ...             -83      4.485           0.083   
7          0.000009  ...             -78      4.457           0.024   
8          0.000016  ...             -89      4.019           0.033   
9          0.000047  ...            -137      4.169           0.055   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   
5          -0.229     0.972          0.315         -0.105  296.28613   
6          -0.028     0.848          0.033         -0.072  286.99948   
7          -0.024     0.964          0.038         -0.038  286.80847   
8          -0.027     1.952          0.099         -0.110  292.24728   
9          -0.045     1.451          0.110         -0.110  281.28812   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  
5  48.224670      15.714  
6  48.375790      15.841  
7  49.316399      11.338  
8  47.969521      10.463  
9  42.451080      13.563  

[10 rows x 41 columns]

# Select features (columns)

In [3]:
# Set features
# selected_features = df[['names', 'of', 'selected', 'features', 'here']]
selected_features = df

# Create a Train Test Split

Use `koi_disposition` for the y values

In [4]:
# Assign x and y values
X = df.drop("koi_disposition", axis=1)
y = df["koi_disposition"]

print(X.shape, y.shape)

(6991, 40) (6991,)


In [5]:
# Import Dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, MinMaxScaler


# Split data into training and testing
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=115, stratify=y)

In [6]:
X_train.head()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  koi_period  \
2700              0              0              0              0   31.804760   
21                0              0              0              0    4.280964   
5040              0              1              1              0    0.829019   
2795              0              0              1              0    1.387838   
1765              0              0              0              0   14.383227   

      koi_period_err1  koi_period_err2  koi_time0bk  koi_time0bk_err1  \
2700         0.000091        -0.000091    190.25413           0.00224   
21           0.000006        -0.000006    171.89659           0.00115   
5040         0.000002        -0.000002    131.46667           0.00401   
2795         0.000006        -0.000006    132.11351           0.00318   
1765         0.000086        -0.000086    140.57648           0.00530   

      koi_time0bk_err2  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
2700          -0.00224  ...            -128      4.235           0.143   
21            -0.00115  ...            -169      4.564           0.032   
5040          -0.00401  ...            -255      4.271           0.124   
2795          -0.00318  ...            -219      4.442           0.054   
1765          -0.00530  ...            -135      4.249           0.137   

      koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
2700          -0.117     1.356          0.212         -0.259  287.62082   
21            -0.168     0.831          0.207         -0.069  294.26581   
5040          -0.186     1.328          0.408         -0.220  296.03485   
2795          -0.216     1.044          0.335         -0.112  295.64212   
1765          -0.112     1.291          0.216         -0.216  290.08276   

            dec  koi_kepmag  
2700  43.141647      13.178  
21    49.314091      15.356  
5040  45.490822      14.661  
2795  48.773071      15.167  
1765  50.863480      13.312  

[5 rows x 40 columns]

# Pre-processing

Scale the data using the MinMaxScaler and perform some feature selection

In [7]:
# Scale the data
X_scaler = MinMaxScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

C:\Users\ofunk\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:334: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by MinMaxScaler.
  return self.partial_fit(X, y)


# Train the Model - Random Forest
Use `Random Forest` model

In [8]:
# Create the Random Forest Model
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier()
rf = rf.fit(X_train_scaled, y_train)

C:\Users\ofunk\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


In [9]:
print(f"Training Data Score: {rf.score(X_train_scaled, y_train)}")
print(f"Testing Data Score: {rf.score(X_test_scaled, y_test)}")

Training Data Score: 0.9954224680526416
Testing Data Score: 0.8792906178489702


In [10]:
# Random Forests in sklearn will automatically calculate feature importance
importances = rf.feature_importances_
importances

array([0.1029767 , 0.05732885, 0.10695514, 0.02592259, 0.02175238,
       0.01798352, 0.0125995 , 0.0109203 , 0.0133002 , 0.0185507 ,
       0.01868529, 0.01276347, 0.0102456 , 0.02393288, 0.03099897,
       0.03479527, 0.03019078, 0.01177764, 0.01166388, 0.03342301,
       0.02904871, 0.0377137 , 0.02033727, 0.00745112, 0.02537639,
       0.00860828, 0.066597  , 0.00251086, 0.01070025, 0.01489173,
       0.07146088, 0.00899603, 0.00951346, 0.0095744 , 0.01085911,
       0.01696198, 0.00881011, 0.01074519, 0.01253272, 0.01054415])

In [11]:
# Sort the features by their importance
sorted(zip(rf.feature_importances_, selected_features), reverse=True)

[(0.10695514089282569, 'koi_fpflag_ss'),
 (0.10297669897116642, 'koi_disposition'),
 (0.07146088113747416, 'koi_steff_err1'),
 (0.06659700217591538, 'koi_insol_err2'),
 (0.05732884816373189, 'koi_fpflag_nt'),
 (0.037713697452225046, 'koi_prad_err1'),
 (0.03479526688571124, 'koi_duration_err1'),
 (0.03342300670013744, 'koi_depth_err2'),
 (0.03099897290148095, 'koi_duration'),
 (0.030190778387271244, 'koi_duration_err2'),
 (0.02904871126083104, 'koi_prad'),
 (0.02592259195417685, 'koi_fpflag_co'),
 (0.025376393368735463, 'koi_insol'),
 (0.02393287518780162, 'koi_impact_err2'),
 (0.021752375281195505, 'koi_fpflag_ec'),
 (0.020337269689877722, 'koi_prad_err2'),
 (0.018685292932792096, 'koi_time0bk_err2'),
 (0.01855069693274302, 'koi_time0bk_err1'),
 (0.01798352298508559, 'koi_period'),
 (0.016961977752312368, 'koi_srad'),
 (0.0148917292652886, 'koi_steff'),
 (0.013300199447768182, 'koi_time0bk'),
 (0.01276346952272514, 'koi_impact'),
 (0.012599503461819095, 'koi_period_err1'),
 (0.01253272

# Hyperparameter tuning 

Use `GridSearchCV` to tune the Random Forest model's parameters

In [12]:
# Create the GridSearchCV model
from sklearn.model_selection import GridSearchCV
param_grid = {'n_estimators': [250, 300, 350],
              'max_depth': [125, 150, 175]}
grid = GridSearchCV(rf, param_grid, verbose=3)

In [13]:
# Fit the Random Forest model using the grid search estimator. 
# This will take the Random Forest model and try each combination of parameters
grid.fit(X_train_scaled, y_train)

Fitting 3 folds for each of 9 candidates, totalling 27 fits
[CV] max_depth=125, n_estimators=250 .................................


C:\Users\ofunk\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.


[CV]  max_depth=125, n_estimators=250, score=0.8890160183066361, total=   2.3s
[CV] max_depth=125, n_estimators=250 .................................


[Parallel(n_jobs=1)]: Done   1 out of   1 | elapsed:    2.4s remaining:    0.0s


[CV]  max_depth=125, n_estimators=250, score=0.8964530892448512, total=   2.3s
[CV] max_depth=125, n_estimators=250 .................................


[Parallel(n_jobs=1)]: Done   2 out of   2 | elapsed:    4.9s remaining:    0.0s


[CV]  max_depth=125, n_estimators=250, score=0.8883800801373783, total=   2.2s
[CV] max_depth=125, n_estimators=300 .................................
[CV]  max_depth=125, n_estimators=300, score=0.8907322654462243, total=   2.9s
[CV] max_depth=125, n_estimators=300 .................................
[CV]  max_depth=125, n_estimators=300, score=0.8953089244851259, total=   2.9s
[CV] max_depth=125, n_estimators=300 .................................
[CV]  max_depth=125, n_estimators=300, score=0.8889524899828277, total=   2.9s
[CV] max_depth=125, n_estimators=350 .................................
[CV]  max_depth=125, n_estimators=350, score=0.8907322654462243, total=   3.3s
[CV] max_depth=125, n_estimators=350 .................................
[CV]  max_depth=125, n_estimators=350, score=0.8953089244851259, total=   3.2s
[CV] max_depth=125, n_estimators=350 .................................
[CV]  max_depth=125, n_estimators=350, score=0.8895248998282771, total=   3.3s
[CV] max_depth=150, n

[Parallel(n_jobs=1)]: Done  27 out of  27 | elapsed:  1.4min finished


GridSearchCV(cv='warn', error_score='raise-deprecating',
       estimator=RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False),
       fit_params=None, iid='warn', n_jobs=None,
       param_grid={'n_estimators': [250, 300, 350], 'max_depth': [125, 150, 175]},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring=None, verbose=3)

In [14]:
# List the best parameters for this dataset
print(f"Best Parameters: {grid.best_params_}")

# List the best score
print(f"Best Score: {grid.best_score_}")

Best Parameters: {'max_depth': 150, 'n_estimators': 250}
Best Score: 0.8930001907304979


In [15]:
print(f"Training Grid Score: {grid.score(X_train_scaled, y_train)}")
print(f"Testing Grid Score: {grid.score(X_test_scaled, y_test)}")

Training Grid Score: 1.0
Testing Grid Score: 0.8895881006864989


In [16]:
# Make predictions with the hypertuned model
predictions = grid.predict(X_test_scaled)

In [17]:
# Calculate classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions))

                precision    recall  f1-score   support

     CANDIDATE       0.82      0.72      0.77       422
     CONFIRMED       0.78      0.84      0.81       450
FALSE POSITIVE       0.98      1.00      0.99       876

     micro avg       0.89      0.89      0.89      1748
     macro avg       0.86      0.85      0.85      1748
  weighted avg       0.89      0.89      0.89      1748



# Save the Model

In [18]:
# Save the model 

import joblib
filename = 'ofunkey_RF.sav'
joblib.dump(rf, filename)

['ofunkey_RF.sav']

# Conclusion


The f1-score of FALSE POSITIVE for Random Forest model is closer to 1 (0.99) meaning it can predict FALSE POSITIVE well and it's reliable.

The Grid score from the hyperparameter tuning boost the Random Forest score higher
Training Data Score: 0.9954224680526416 changes to 1
Testing Data Score:  0.8792906178489702 changes to 0.8895881006864989
Best Score: 0.8930001907304979 

However, Random Forest model's best score of (0.8930001907304979) seems better than SVM model (0.8706847224871257) when comparing the scores.

